flytekit/core/type_engine.py

```diff
@@ -793,6 +826,98 @@ def guess_python_type(self, literal_type: LiteralType) -> Type[list]:
             return typing.List[ct]
         raise ValueError(f"List transformer cannot reverse {literal_type}")
 
+def _add_tag_to_type(x: LiteralType, tag: str) -> LiteralType:
+    x._structure = TypeStructure(tag=tag)
+    return x
+
+class UnionTransformer(TypeTransformer[T]):
+    """
+    Transformer that handles a typing.Union[T1, T2, ...]
+    """
+
+    def __init__(self):
+        super().__init__("Typed Union", typing.Union)
+
+    def get_literal_type(self, t: Type[T]) -> Optional[LiteralType]:
+        try:
+            trans = [(TypeEngine.get_transformer(x), x) for x in _get_args(t)]
+            variants = [_add_tag_to_type(t.get_literal_type(x), t.name) for (t, x) in trans]
+            return _type_models.LiteralType(union_type=UnionType(variants))
+        except Exception as e:
+            raise ValueError(f"Type of Generic Union type is not supported, {e}")
+
+    def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], expected: LiteralType) -> Literal:
+        found_res = False
+        res = None
+        res_type = None
+        for t in _get_args(python_type):
+            try:
+                trans = TypeEngine.get_transformer(t)
+
+                res = trans.to_literal(ctx, python_val, t, expected)
+                res_type = _add_tag_to_type(trans.get_literal_type(t), trans.name)
+                if found_res:
+                    # Should really never happen, sanity check
+                    raise TypeError(f"Ambiguous choice of variant for union type")
+                found_res = True
+            except TypeTransformerFailedError as e:
+                logger.debug(f"Failed to convert from {python_val} to {t}", e)
+                continue
+
+        if found_res:
+            return Literal(scalar=Scalar(union=Union(value=res, stored_type=res_type)))
+
+        raise TypeTransformerFailedError(f"Cannot convert from {python_val} to {python_type}")
+
+    def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type: Type[T]) -> Optional[typing.Any]:
+        union_tag = None
+        if lv.scalar is not None and lv.scalar.union  is not None:
+            union_type = lv.scalar.union.stored_type
+            if union_type.structure is not None:
+                union_tag = union_type.structure.tag
+
+        found_res = False
+        res = None
+        res_tag = None
+        for v in _get_args(expected_python_type):
+            try:
+                trans = TypeEngine.get_transformer(v)
+                if union_tag is not None:
+                    if trans.name != union_tag:
+                        continue
```

Comment:
some notes, for the record:

a value of type `List[int]` should be assignable to a value of type `Union[List[Union[int, str]], List[bool]]` since the types `List[int]` and `List[Union[int, str]]` are compatible
this means that the stored type of the literal does not necessarily match either of the union variants 1-for-1

instead the types must be "castable" (see `AreTypesCastable` from propeller) or, really, the `Literal` must be compatible with the type transformer for the target type

the only way to check whether a type transformer accepts a `Literal` is to try running it and look for exceptions (generic exception previously, `TypeTransformerFailedError` in this PR). this is exactly what the current implementation does

I wil implement `AreTypesCastable` in flytekit and rely on it instead. note that this introduces a discrepancy between whether a type transformer is supposed to succeed and whether it actually succeeds, but I think it's fine because I cannot come up with a counter-example. my reasoning relies on the failure of a type transformer that matches the literal type being non-fatal i.e. causing a `continue` instead of an exception in the union type transformer loop

the main dangerous interaction here is that the `LiteralType` -> `TypeTransformer` mapping is non-injective. what saves us, however, is that the Python Type -> `TypeTransformer` mapping *is* injective and, as a consequence, the `[TypeEngine.to](http://TypeEngine.to)_literal_type(x: Type)` function is also injective. if it was not injective, comparing `LiteralType`s would be impossible because we would not have enough information to get the actual expected `LiteralType`